torch/_inductor/fx_passes/pre_grad.py

```diff
@@ -340,3 +344,59 @@ def transpose_matmul(A: torch.Tensor, B: torch.Tensor, Atrans: bool, Btrans: boo
     if Btrans:
         B = B.transpose(-1, -2)
     return torch.matmul(A, B)
+
+
+_pointwise_ops = [
+    operator.add,
+    operator.sub,
+    operator.mul,
+    operator.iadd,
+    operator.isub,
+    operator.imul,
+    operator.itruediv,
+    operator.ifloordiv,
+]
+
+
+def boolean_mask_rewritten(graph: torch.fx.Graph) -> torch.fx.Graph:
+    for node in graph.nodes:
+        if CallFunctionVarArgs(operator.setitem).match(node):
+            # value = pointwise_op(tensor[index])
+            # tensor[index] = value
+            tensor, index, value = node.args
+            if (
+                isinstance(index, torch.fx.node.Node)
+                and "example_value" in index.meta
+                and isinstance(index.meta["example_value"], FakeTensor)
+                and index.meta["example_value"].dtype == torch.bool
+            ):
+                if CallFunctionVarArgs(_pointwise_ops).match(value):
+                    pointwise_op = value.target
+                    arg1, arg2 = value.args
+                    if CallFunctionVarArgs(operator.getitem).match(arg1):
+                        _tensor, _index = arg1.args
+                        if _tensor is tensor and _index is index:
```

Comment:
@SherlockNoMad Is this the best we can do for writing a pattern match like this?

Reply:
Discussed with Yanbo offline. Subgraph pattern matcher should be able to help. 
But will need to reland the diff that supports wildcard matching. 
In this pattern, it's 3 node
`get_item`, `oneof(elementwise ops)`, `set_item`

Reply:
It seems Subgraph pattern matcher doesn't work:
```
def boolean_mask_rewritten(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:

    def pattern(x, y, z):
        a = operator.getitem(x, y)
        b = operator.imul(a, z)
        operator.setitem(x, y, b)
        return x

    def replacement(x, y, z):
        x = torch.where(x, y * z, x)
        return x

    torch.fx.subgraph_rewriter.replace_pattern(gm, pattern, replacement)
    gm.graph.lint()
```
Error:
```
TypeError: 'Proxy' object does not support item assignment
```
Seems symbolic tracing doesn't support ```setitem```.

Reply:
I would not hard block on using the nicer API. In this case it seems that it would be easier if we could normalize the IR first. But that requires pre dispatch tracing ;) @bdhirsh   Or maybe this pattern match could be done post AOTAutograd? Choices choices.

Reply:
I talked to Horace and Yanbo a few weeks ago - pre-dispatch tracing is ready to use, minus the fact that it doesn't have massive amounts of testing yet. If someone from the inductor team is ready to use it, I'm 100% happy to help jump on and sift through any bugs that come out.

One known issue today is that it doesn't handle higher order ops, which @tugsbayasgalan has a fix for: https://github.com/pytorch/pytorch/pull/107885